packages/text-annotator/src/highlight/canvas/canvasRenderer.ts

```diff
@@ -0,0 +1,121 @@
+import type { HighlightStyle } from '../HighlightStyle';
+import { DEFAULT_SELECTED_STYLE, DEFAULT_STYLE, HighlightStyleExpression } from '../HighlightStyle';
+import type { HighlightPainter } from '../HighlightPainter';
+import { createBaseRenderer, type RendererImplementation } from '../baseRenderer';
+import type { Highlight } from '../Highlight';
+import type { TextAnnotatorState } from 'src/state';
+import type { ViewportState } from '@annotorious/core';
+
+const createCanvas = () => {
+  const canvas = document.createElement('canvas');
+  canvas.width = window.innerWidth;
+  canvas.height = window.innerHeight;
+  canvas.className = 'r6o-highlight-layer bg';
+  return canvas;
+}
+
+const resetCanvas = (canvas: HTMLCanvasElement, highres?: boolean) => {
+  canvas.width = highres ? 2 * window.innerWidth : window.innerWidth;
+  canvas.height = highres ? 2 * window.innerHeight : window.innerHeight;
+
+  if (highres) {
+    // Note that resizing the canvas resets the context
+    const context = canvas.getContext('2d');
+    context.scale(2, 2);
+    context.translate(0.5, 0.5);
+  }
+}
+
+const createRenderer = (container: HTMLElement): RendererImplementation => {
+
+  container.classList.add('r6o-annotatable');
+
+  const canvas = createCanvas();
+  const ctx = canvas.getContext('2d');
+
+  container.insertBefore(canvas, container.firstChild);
+
+  const redraw = (   
+    highlights: Highlight[], 
+    viewportBounds: ViewportBounds,
+    currentStyle?: HighlightStyleExpression,
+    painter?: HighlightPainter
+  ) => requestAnimationFrame(() => {
+
+    const { width, height } = canvas;
+
+    // New render loop - clear canvases
+    ctx.clearRect(-0.5, -0.5, width + 1, height + 1);
+
+    if (painter)
+      painter.clear();
+
+    const { top, left } = viewportBounds;
+    
+    highlights.forEach(h => {
+      const base: HighlightStyle = currentStyle 
+        ? typeof currentStyle === 'function' 
+          ? currentStyle(h.annotation, h.state) 
+          : currentStyle 
+        : h.state?.selected 
+          ? DEFAULT_SELECTED_STYLE 
+          : DEFAULT_STYLE;
+
+      // Trigger the custom painter (if any) as a side-effect
+      const style = painter ? painter.paint(h, viewportBounds) || base : base;
+
+      // Offset annotation rects by current scroll position
+      const offsetRects = h.rects.map(({ x, y, width, height }) => ({ 
+        x: x + left, 
+        y: y + top, 
+        width, 
+        height 
+      }));
+
+      ctx.fillStyle = style.fill;
+      ctx.globalAlpha = style.fillOpacity || 1;
+      
+      offsetRects.forEach(({ x, y, width, height }) => ctx.fillRect(x, y - 2.5, width, height + 5));
+
+      if (style.underlineColor) {
+        ctx.globalAlpha = 1;
+        ctx.strokeStyle = style.underlineColor;
+
+        offsetRects.forEach(({ x, y, width, height }) => {
+          ctx.beginPath();
+          ctx.moveTo(x, y + height + 4);
+          ctx.lineTo(x + width, y + height + 4);
```

Comment:
I was about to suggest a change that would allow using the `underlineThickness` & `underlineOffset` for the `CANVAS` renderer:
![image](https://github.com/recogito/text-annotator-js/assets/68850090/82bc5301-1c82-40e3-a337-70c6205f01e5)

But unfortunately, I'm not sure why the `4` was chosen as a default offset value here 🤔. Should the custom offset override that `4` or should it sum with it?

Reply:
I _suppose_ that it's _the latter_. Because if I set the offset to `0`, the underline will render over the highlight:
![image](https://github.com/recogito/text-annotator-js/assets/68850090/b57b1131-1918-4e4a-9009-839a0fea354e)

Reply:
Hi @oleksandr-danylchenko, I think the Canvas renderer might simply have gotten stuck in mid-development, since I currently switched our client implementation entirely to the SPAN renderer. You are right, of course, that the offset should be configurable, and 0 should align with the box.

I don't have the details in the back of my head exactly, but it's a tricky problem. The annotation highlight size is definitely influenced by various factors, most importantly the CSS `line-height`, and differences between Chrome and FF. However, once the box size is determined and the Canvas renderer uses `ctx.rect()` to draw the box, you should also be able to draw a line that aligns with it. It may not align 100% with the browser selection highlight, but should definitely align with the box drawn on the canvas. Is this not what's happening?

Reply:
*Added a PR for the `underlineThickness` & `underlineOffset` processing - https://github.com/recogito/text-annotator-js/pull/89

Reply:
> It may not align 100% with the browser selection highlight

Fortunately, I made the browser's and annotator highlights work separately in my app. The latter will appear only once the selection is finished. So it should be enough to depend only on the `ctx.rect()`.

> However, once the box size is determined and the Canvas renderer uses ctx.rect() to draw the box, you should also be able to draw a line that aligns with it

I'll play with the highlight dimensions a bit to whether the rects sizes can be programmatically synced. So we won't use the "magic" `4`
![image](https://github.com/recogito/text-annotator-js/assets/68850090/4cf43f2e-d0a0-4203-87bb-2fb921002491)